### PR TITLE
Atomic Batch Update ASPA Records in ASPA Table

### PR DESCRIPTION
**Pull request by carl-tud**
# Work in Progress

## Changes related to storing ASPA records
This PR adds atomic update functionality to the `aspa_table`. If an update fails in the rtr_sync loop, changes must be reverted. Changes made due to withdrawn records however cannot be undone as withdrawals are inherently lossy (provider ASNs are purged at withdrawal, so cannot be restored upon "undo update").

- This PR introduces a new `aspa_update_operation` struct: Models an "**add record**" or **remove record** operation.
- `aspa_table_compute_update` generates a new `aspa_array` given an array of add/remove operations. This function *does not* make any modifications to the ASPA table supplied. The new `aspa_array` is stored in a new struct `aspa_update` which holds additional information on provider ASN array that need be released in case the update gets applied by the caller. By encapsulating the results of an update operation into a separate struct the update can be discarded if another function invoked by the caller fails after `aspa_table_compute_update` has been called. This mechanism cleverly avoids having us to implement an "**undo update**" algorithm that would need to reverse changes already written to memory once a function fails that's being executed after the update to the ASPA table would have already been completed.
- `aspa_table_apply_update` switches the ASPA array in the table for the new one and frees old provider ASN arrays (which were references by now-deleted ASPA records inside the old ASPA array).

### CODE REVIEW

**Comment by carl-tud** [comment]
I think we should move the remaining aspa_array functions into aspa.c (as static funds) and aspa_private.h respectively.

File: rtrlib/aspa/aspa.c [hunk 1]
```diff
@@ -43,38 +48,77 @@ RTRLIB_EXPORT void aspa_table_init(struct aspa_table *aspa_table, aspa_update_fp
 	pthread_rwlock_init(&(aspa_table->lock), NULL);
 }
 
+static enum aspa_rtvals aspa_table_remove_node(struct aspa_table *aspa_table, struct aspa_store_node **node, bool notify)
+{
+	if (!node || *node) {
```
Comment (moversity) [inline rtrlib/aspa/aspa.c]: why's it a problem if `*node != NULL`?

File: rtrlib/aspa/aspa.c [hunk 2]
```diff
@@ -194,42 +238,285 @@ RTRLIB_EXPORT int aspa_table_remove(struct aspa_table *aspa_table,
...
+		for (size_t i = 0; i < existing_array->size; i++) {
+			aspa_table_notify_clients(aspa_table, &(existing_array->data[i]), update->rtr_socket, false);
+		}
+		
+		// Free the old array
+		aspa_array_free(existing_array);
+		
+		// Notify clients new records have been added
+		for (size_t i = 0; i < update->aspa_array->size; i++)
+			aspa_table_notify_clients(aspa_table, &( update->aspa_array->data[i]), update->rtr_socket, true);
```
Comment (moversity) [inline rtrlib/aspa/aspa.c]: shouldn't this only be done for what's actually changed?

File: rtrlib/aspa/aspa.c [hunk 3]
```diff
@@ -194,42 +238,285 @@ RTRLIB_EXPORT int aspa_table_remove(struct aspa_table *aspa_table,
...
+        aspa_array_free(update->aspa_array);
+	
+	if (update->stale_provider_arrays && update->stale_provider_arrays_count > 0) {
+		for (size_t i = 0; i < update->stale_provider_arrays_count; i++)
+			lrtr_free(update->stale_provider_arrays[i]);
+		
+		lrtr_free(update->stale_provider_arrays);
+	}
+	
+    lrtr_free(update);
```
Comment (moversity) [inline rtrlib/aspa/aspa.c]: either fix indentation or brackets

Comment (moversity) [reply rtrlib/aspa/aspa.c]: oh, i just noticed, can you please change `    ` for `\t`?

File: rtrlib/aspa/aspa.c [hunk 4]
```diff
@@ -194,42 +238,285 @@ RTRLIB_EXPORT int aspa_table_remove(struct aspa_table *aspa_table,
...
+    // perform lookup on aspa_table->store insted.
+
+	struct aspa_array **array = NULL;
+    // Use fast lookup
+    if (rtr_socket->aspa_table == aspa_table && rtr_socket->aspa_array) {
+        // Check if an ASPA array exists for this socket
+		array = &rtr_socket->aspa_array;
+    }
+	
+	if (!array) {
```
Comment (moversity) [inline rtrlib/aspa/aspa.c]: this could also be an else clause to the previous if

**Pull request closed without merging**